### PR TITLE
More CUDA intrinsics

### PR DESCRIPTION
**Pull request by jdtatz**
Added support for the warp intrinsics, two more special registers, and the other sync-threads functions. I tested everything on a 1080Ti at work, but I don't have accesses to a Volta GPU to test its new features (matching and independent scheduling).

TODO:
- [x] Add documentation
- [x] Add wrappers around the shfl_sync and vote_sync intrinsics to make them look more like the normally used functions
- [x] Add more thorough tests
- [ ] Support for warps in the CUDA simulator. Unsure how to do this correctly.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
TODO: Before merge run this through the buildfarm CUDA test machines.

File: numba/cuda/intrinsic_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import numba
+
+
+@jit(device=True)
+def all_sync(mask, predicate):
+    """
+    If for all threads in the masked warp the predicate is true, then
+    a non-zero value is returned, otherwise 0 is returned.
+    """
+    return numba.cuda.vote_sync_intrinsic(mask, 0, predicate)[1]
```
[the remark this answers is not in the document]
Comment (jdtatz) [reply numba/cuda/intrinsic_wrapper.py]: enum isn't supported in the CUDA jit.

**Comment by stuartarchibald** [comment]
@jdtatz thanks for the fix ups, looks good. Pending success on the buildfarm this can be merged.

**Comment by stuartarchibald** [comment]
Build is failing on the build farm for combination `Python 3.6`, `NumPy 1.14`, `CUDA 9.0` on Windows 64 bit. Error is:
```
ERROR: test_vote_sync_ballot (numba.cuda.tests.cudapy.test_warp_ops.TestCudaWarpOperations)
Traceback:
File "<path>"\numba\cuda\tests\cudapy\test_warp_ops.py", line 208, in test_vote_sync_ballot
    self.assertTrue(np.all(ary = np.int32(0xffffffff)))
OverflowError: Python int too large to convert to C long
```
I'm guessing this is windows having default int as 32bit and linux as 64bit.

**Comment by stuartarchibald** [comment]
Thanks for the fix in https://github.com/numba/numba/pull/2910/commits/d0725fa6db086b6fdc05cae3d490b35ed609ff6d, passes on `{python 3.6, python 2.7} x {np 1.14} x {CUDA 8, CUDA 9, CUDA 9.1}`. I think this is sufficient for merge. Thanks again for your contribution.

**Pull request merged**